src/fs.ts

```diff
@@ -34,45 +34,54 @@ export class RemoteFileSystem implements FileSystem {
 	 */
-	async getWorkspaceFiles(base?: string, childOf = new Span()): Promise<Iterable<string>> {
-		return iterate(await this.client.workspaceXfiles({ base }, childOf))
-			.map(textDocument => textDocument.uri);
+	async getWorkspaceFiles(base?: URL, childOf = new Span()): Promise<Iterable<URL>> {
+		return iterate(await this.client.workspaceXfiles({ base: base && base.href }, childOf))
+			.map(textDocument => new URL(textDocument.uri));
 	}
 
 	/**
 	 * The content request is sent from the server to the client to request the current content of any text document. This allows language servers to operate without accessing the file system directly.
 	 */
-	async getTextDocumentContent(uri: string, childOf = new Span()): Promise<string> {
-		const textDocument = await this.client.textDocumentXcontent({ textDocument: { uri } }, childOf);
+	async getTextDocumentContent(uri: URL, childOf = new Span()): Promise<string> {
+		const textDocument = await this.client.textDocumentXcontent({ textDocument: { uri: uri.href } }, childOf);
 		return textDocument.text;
 	}
 }
 
+/**
+ * FileSystem implementation that reads from the local disk
+ */
 export class LocalFileSystem implements FileSystem {
 
 	/**
-	 * @param rootPath The root directory path that relative URIs should be resolved to
+	 * @param rootUri The workspace root URI that is used when no base is given
 	 */
-	constructor(private rootPath: string) {}
+	constructor(protected rootUri: URL) {}
 
 	/**
-	 * Converts the URI to an absolute path
+	 * Returns the file path where a given URI should be located on disk
 	 */
-	protected resolveUriToPath(uri: string): string {
-		return toUnixPath(path.resolve(this.rootPath, uri2path(uri)));
+	protected resolveUriToPath(uri: URL): string {
+		return uri2path(uri);
 	}
 
-	async getWorkspaceFiles(base?: string): Promise<Iterable<string>> {
-		// Even if no base provided, still need to call resolveUriToPath which may be overridden
-		const root = this.resolveUriToPath(base || path2uri('', this.rootPath));
-		const baseUri = path2uri('', normalizeDir(root)) + '/';
+	async getWorkspaceFiles(base: URL = this.rootUri): Promise<Iterable<URL>> {
 		const files = await new Promise<string[]>((resolve, reject) => {
-			glob('*', { cwd: root, nodir: true, matchBase: true }, (err, matches) => err ? reject(err) : resolve(matches));
+			glob('*', {
+				// Search the base directory
+				cwd: this.resolveUriToPath(base),
+				// Don't return directories
+				nodir: true,
+				// Search directories recursively
+				matchBase: true,
+				// Return absolute file paths
+				absolute: true
+			} as any, (err, matches) => err ? reject(err) : resolve(matches));
 		});
-		return iterate(files).map(file => baseUri + file.split('/').map(encodeURIComponent).join('/'));
+		return iterate(files).map(file => new URL(file.split('/').map(encodeURIComponent).join('/'), base.href + '/'));
```

Comment:
Does it handle special characters properly? I see the following example in https://url.spec.whatwg.org/#url:

```
var input = "https://example.org/💩",
    url = new URL(input)
url.pathname // "/%F0%9F%92%A9"
```

There input is non-encoded